MultiSource/Benchmarks/Ptrdist/anagram/anagram.c

```diff
@@ -377,41 +433,64 @@ NewWord(void) {
  * We would normally just use printf, but the string being printed is
  * is a huge pointer (on an IBM PC), so special care must be taken.
  */
-void wprint(char * pch) {
-    printf("%s ", pch);
+void wprint(_Array_ptr<char> pch) {
+    // CHECKEDC : bounds-safe interface is not given
+    printf("%s ", (const char*)pch);
 }
 
 PWord NextWord(void);
 
 /* NextWord -- get another candidate entry, creating if necessary */
 PWord NextWord(void) {
-    PWord pw;
+    // CHECKEDC : automatic checkedc pointer variable initializer
+    PWord pw = 0;
     if (cpwCand >= MAXCAND)
 	Fatal("Too many candidates\n", 0);
+    // CHECKEDC : checkedc array pointer array subscript, in-bounds check
+    // dynamic_check((cpwCand >= 0) && (cpwCand < MAXCAND))
+    // : above reasoning facts can remove redundant dynamic_check insertion
     pw = apwCand[cpwCand++];
     if (pw != NULL)
 	return pw;
+    // CHECKEDC : checkedc array pointer access, in-bounds check
+    // dynamic_check((cpwCand-1) >= 0 && ((cpwCand-1) < MAXCAND))
+    // : reaonsing facts can remove redundant bounds check insertion
     apwCand[cpwCand-1] = NewWord();
     return apwCand[cpwCand-1];
 }
 
 /* BuildWord -- build a Word structure from an ASCII word
  * If the word does not fit, then do nothing.
  */
-void BuildWord(char * pchWord) {
-    unsigned char cchFrequency[ALPHABET];
+// CHECKEDC : function call formal parameters, checkedc array pointer
+// : pchWord declared bounds is from pchDictionary : count(MAXWORDS)
+// TODO
+// - pchWord is pointer to an word but the size of each word is not given by caller
+// - However, caller knows about size of each word
+// - if it is possible to rewrite function, we can use word size information in declaring bounds
+void BuildWord(_Array_ptr<char> pchWord : bounds(pchDictionary, pchDictionary+MAXWORDS)) {
```

Comment:
It is problematic to use pchDictionary as the bounds for pchWord.    pchDictionary is variable sized (as I mentioned earlier).   The problem is at function calls, we must assume that pchDictionary and its length might be changed.  There are calls to bzero in the body of this method.

I think a good solution is to add additional parameters to BuildWord that hold the bounds for pchWord.